Adapt for PHP 7.0

XedinUnknown opened the pull request:
This branch was originally meant to adapt the repo for PHP 7.1. However, it later became apparent that many projects which will rely on this standard in fact support **PHP 7.0**.

- Removed `object` typehint.
- Changed `iterable` typehint to `Traversable`.
- Downgraded PHP version requirement.
- Renamed `StoppableEventInterface` to `IsPropagationStoppedCapableInterface`.

Also:

- PHPUnit config and tests - to ascertain that syntax is compatible.
- Travis config - to test against target PHP versions.
- Docker config - develop and test in an isolated environment.
Copy `.env.example` to `.env`, and run `docker-compose up`.
- Lockfile now committed.
- Configuration for dist building with `git archive`. Results in smaller dist packages.

XedinUnknown committed:
Package now requires PHP 7.0

This is a downgrade from 7.2.

XedinUnknown committed:
Removed PHP 7.2 syntax

- Removed `object` typehint.
- Changed `iterable` typehint to `Traversable`.

XedinUnknown committed:
Lockfile now allowed in repo

XedinUnknown committed:
Added PHPUnit

XedinUnknown committed:
Added PHPUnit configuration

XedinUnknown committed:
Added tests

XedinUnknown committed:
Added Docker environment configuration

XedinUnknown committed:
Exclude generated test data

XedinUnknown committed:
Add PHPStorm configuration

XedinUnknown committed:
Add Travis config

XedinUnknown committed:
Add configuration for dist building with git archive

XedinUnknown committed:
Amend readme

Now explains why the package exists, and what has changed.

XedinUnknown committed:
Updated CI to test on PHP 7.4

It has been released, and therefore it looks like the snapshot version is no longer available.

XedinUnknown committed:
Switch to Xenial distro for CI

Travis is unable to install PHP 7.4. Perhaps it is because Precise is too old.

mecha commented on test/bootstrap.php:
```diff
@@ -0,0 +1,10 @@
+    if (file_exists($autoload)) {
+        require_once $autoload;
+    }
+}, [__FILE__]);
```
Interesting approach. Any reason for enclosing in a function?

Also I think you forgot to consume the argument, or forgot to remove it.

XedinUnknown replied:
It's an [IIFE][]. Here, I use it to avoid polluting the global variable scope.

But yes, I forgot to use the argument. Fixing.

[IIFE]: https://en.wikipedia.org/wiki/Immediately_invoked_function_expression

mecha commented on src/ListenerProviderInterface.php:
```diff
@@ -11,9 +13,9 @@ interface ListenerProviderInterface
      *   callable MUST be type-compatible with $event.
      */
-    public function getListenersForEvent(object $event) : iterable;
+    public function getListenersForEvent($event) : Traversable;
```
Arrays do not satisfy `Traversable`, so this fork is strictly speaking not API-compatible. I suggest removing the return type hint altogether.

XedinUnknown replied:
Arrays do not satisfy `Traversable`, but `Traversable` satisfies `iterable`. Because earlier PHP versions do not have the `iterable` type, I had to choose between `array` and `Traversable`. I chose `Traversable` because it allows for more flexibility.

The premise for forward-compatibility here is that all implementations of this fork would still work just the same if they suddenly begun to also implement the original PSR-14 interfaces. Doesn't this interface satisfy that condition?

mecha replied:
In that case, yes. If FC is what's important then `Traversable` is the logical choice.

But in the event of porting an implementation over to this fork, `array` returns would need to be wrapped. It would probably be a good idea to make a list somewhere, like in the README, that outlines what changes would need to be made to an implementation to port it over to this forked standard.

XedinUnknown replied:
That's true. But I haven't considered that someone may have a need to port PSR-14 implementations to this. I thought that mostly people would be porting from this to PSR-14 :D. But you're right!

XedinUnknown replied:
Done.

XedinUnknown committed:
Fix not using provided root file path in bootstrap

XedinUnknown committed:
Allow builds on PHP 7.4 to fail

XedinUnknown committed:
Updated dependencies

XedinUnknown committed:
Changed ownership

- Changed namespace from `Psr` to `Dhii`.
- Changed package name.
- Updated Readme.
- Updated autoloading config.

XedinUnknown committed:
Fixed Docker Compose config version

Now matches that supported by PHPStorm.

XedinUnknown committed:
Merge branch 'develop' into task/adapt-for-php-7.1

XedinUnknown committed:
Changed namespace structure

- Dispatcher, event, and listener interfaces now live in their own namespace.
- Root namespace is now `Dhii\Events`. The term `EventDispatcher` is too specific.

XedinUnknown committed:
Merge branch 'develop' into task/adapt-for-php-7.1

XedinUnknown committed:
Renamed project

XedinUnknown committed:
Change branch alias for `1.0.x-dev` from `master` to `develop`

XedinUnknown committed:
Add porting instructions to and from PSR-14

mecha commented on README.md:
```diff
@@ -25,4 +25,36 @@ Copy `.env.example` to `.env`, and run `docker-compose up`.
+The reason is that, while `Traversable` is a subtype of `iterable`, `array` is not a subtype of
+`Traversable`. Therefore, if your `ListenerProviderInterface` implementation returns arrays
+(which is valid with PSR-14), it must start returning `Traversable` instances instead. The simplest
+way to do this is by wrapping the array into an [`ArrayObject`][].
```
I think [`ArrayIterator`](https://www.php.net/manual/en/class.arrayiterator.php) would be a better suggestion here.

XedinUnknown replied:
Could you elaborate on why that is so?

mecha replied:
The result should be a `Traversable`, and `ArrayIterator` simply wraps an array into an `Iterator`, making it the ideal minimal requirement. `ArrayObject` on the other hand implements a whole bunch of interfaces. Not to mention that it's technically not an `Iterator`, but an `IteratorAggregate`, therefore imposing a small overhead during iteration.

XedinUnknown replied:
![image](https://user-images.githubusercontent.com/1428973/72205238-035be100-3481-11ea-91dc-95d650bb1195.png)
Doesn't look so minimal to me.

mecha replied:
Wow I'm surprised. I guess I should have opened my own link before pasting :man_facepalming:

Though I still think there's an argument to be made here, as conceptually `ArrayObject` wraps arrays into an object while `ArrayIterator` into an `Iterator`. But I won't push my case anymore. PHP doing me a disappoint.

XedinUnknown replied:
I totally get the point you are trying to make. It sounds like it's just a iterator, and nothing more. But in reality, the only difference is the ability to seek, apparently.

XedinUnknown committed:
Renamed `StoppableEventInterface` to `IsPropagationStoppedCapableInterface`

XedinUnknown committed:
Update PHPStorm config

- Removed useless Docker service.
- Removed useless PHP interpreter.
- Tests should now use default interpreter by default.
- The `test` folder is not test source root.

mecha commented on src/Event/IsPropagationStoppedCapableInterface.php:
```diff
@@ -10,7 +10,7 @@
  * return immediately without calling any further Listeners.
  */
-interface StoppableEventInterface
+interface IsPropagationStoppedCapableInterface
```
Why rename? I know that this name makes more sense, but I thought that the purpose of this fork was to support PHP 7.0. I don't see why the name needs to change. It just adds more distance between this fork and the original.

XedinUnknown replied:
I have no problem with distance, since there's no way changes from here will be merged to original PSR-14. The real reason, however, is that I needed to make an actual `StoppableEventInterface`, which extends an actual `EventInterface`. Didz me a conflict.

mecha replied:
Ah, understood.

XedinUnknown commented:
@mecha, do you think I should add dispatcher and listener-provider factories here?

XedinUnknown committed:
Updated readme

Specifically, mentioned `StoppableEventInterface` rename.

XedinUnknown committed:
Updated PHPStorm config

mecha commented:
No

XedinUnknown committed:
Add `AddListenerCapableInterface`

XedinUnknown committed:
Add `RemoveListenerCapableInterface`
